Subject: Don't fail silently when MTA is not installed 
Author: Christian Kastner <debian@kvr.at>
Last-Update: 2010-04-22

When an MTA is not installed (MAILCMD is not available), the attempted piping
out job output to the MTA fails fatally when the pipe's buffer is exceeded.
This patch checks if MAILCMD is available before proceeding to pipe output to
it.

Bug-Debian: http://bugs.debian.org/577133
Index: patched/do_command.c
===================================================================
--- patched.orig/do_command.c	2010-05-06 18:18:59.066435924 +0200
+++ patched/do_command.c	2010-05-06 18:18:59.713435868 +0200
@@ -23,6 +23,8 @@
 #include "cron.h"
 #include <signal.h>
 #include <grp.h>
+#include <sys/stat.h>
+#include <unistd.h>
 #if defined(sequent)
 # include <sys/universe.h>
 #endif
@@ -492,6 +494,8 @@
 // status.
 	
 	long pos;
+	struct stat	mcsb;
+	int		statret;	
 
 	fseek(tmpout, 0, SEEK_END);
 	pos = ftell(tmpout);
@@ -508,6 +512,17 @@
 	else if (!*mailto)
 		mailto = NULL;
 
+	/* Don't send mail if MAILCMD is not available */
+	if ((statret = stat(MAILCMD, &mcsb)) != 0) {
+		Debug(DPROC|DEXT, ("%s not found, not sending mail\n", MAILCMD))
+		if (pos > 0) {
+			log_it("CRON", getpid(), "info", "No MTA installed, discarding output");
+		}
+		goto mail_finished;
+	} else {
+		Debug(DPROC|DEXT, ("%s found, will send mail\n", MAILCMD))
+	}
+
 	register FILE	*mail = NULL;
 	register int	bytes = 1;
 
@@ -617,6 +632,7 @@
 		log_it(usernm, getpid(), "MAIL", "stream error reading output");
 	}
 
+mail_finished:
 	fclose(tmpout);
 
 	if (log_level >= 2) {
